### PR TITLE
Various fixes and improvements to the API docs / Swagger specs

### LINKED ISSUE #31386: Undocumented field in `GET /info` response
**Description**

The Docker API reference https://docs.docker.com/engine/api/v1.26/ does not document the `Swarm` field which is returned in a response of `GET /v1.26/info`. In addition, the field is included in [the Go client type definitions](https://godoc.org/github.com/docker/docker/api/types#Info).

Is it safe to use the undocumented field in client applications?

### PR DESCRIPTION
**Pull request by thaJeztah**
- fixes https://github.com/moby/moby/issues/31386
- follow-up to https://github.com/moby/moby/pull/34478

This builds on top of https://github.com/moby/moby/pull/34478, so needs a rebase after that is merged.

**- What I did**

Fixed various issues in the API documentation / Swagger:

- Added more per-field example values, so that the examples can more easily be reused. When providing a full example _response_, the type/schema of the endpoint is fully ignored by the generated documentation. Because of this, the documentation may _look_ correct (API response looks correct), but the actual definition is incorrect. It also causes a lot of duplication (same example responses used for multiple endpoints).
- Updated / added descriptions for some fields.
- Fixed various incorrect types (e.g. `uint64` instead of `int`);
  - `ObjectVersion.Index` is an `uint64` https://github.com/moby/moby/blob/0fd90c4d5d8ab739d6f999a36d9c4096e89c8059/api/types/swarm/common.go#L5-L15
  - `ClusterInfo` is nullable in the `/info` output (see https://github.com/moby/moby/blob/ff4f700f74450018f36d014f3cde0ff1b9c17fb3/api/types/swarm/swarm.go#L203)
  - `CAConfig.ForceRotate` was missing a type, therefore treated as an `object` in Swagger: https://github.com/moby/moby/blob/ff4f700f74450018f36d014f3cde0ff1b9c17fb3/api/types/swarm/swarm.go#L121
  - `Raft.SnapshotInterval`, `Raft.KeepOldSnapshots`, and `Raft.LogEntriesForSlowFollowers` are an `uint64` not an `int64`
  - Various fields in `swarm.Info` are nullable; added `x-nullable`

- Fix TLSInfo documenting the wrong type, and being at the wrong nesting level;
  - `TLSInfo` is part of `Node.Description`, but was documented as a
direct child of `Node`
  - `Node.TLSInfo` incorrectly was using the `SwarmSpec` type,
instead of `TLSInfo`

- Add definitions for `Node.Status` and `Node.ManagerStatus`, which only present in the response example, but not documented
- Extracted `Platform`, `EngineDescription`, `NodeDescription` to a definition, instead of being defined as part of an API endpoint (allowing them to be re-used).
- Create a definition for the `/info` response (`SystemInfo`); This endpoint was badly documented, missing various fields, and incorrectly describing others. Also removing the response example in favor of per-field examples, as this prevents an incorrectly formatted response from masking omissions in the actual type.

**- How I did it**

Editing, reading the Go structs that are used for the response, and writing/updating the Swagger specs based on that.

**- How to verify it**

run `make swagger-docs` and visit http://localhost:9000 in your browser

**- Description for the changelog**

    Various fixes and improvements to the API documentation /  Swagger specification.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
Ping @vdemeester @dnephin @mstanleyjones PTAL 😅

**Comment by thaJeztah** [comment]
I kept commits separate, so it may be easier to review per commit, to verify the changes; the first commit is https://github.com/moby/moby/pull/34478 so can either be ignored here, or we can close the other one :smile:

File: api/swagger.yaml [hunk 1]
```diff
@@ -868,26 +868,224 @@ definitions:
...
+        - HostIp: "0.0.0.0"
+          HostPort: "80"
+      "53/udp":
+        - HostIp: "0.0.0.0"
+          HostPort: "53"
+      "2377/tcp": null
+
+  PortBinding:
+    description: |
+      PortBinding represents a binding between a Host IP address and a Host
```
Comment (mstanleyjones) [inline api/swagger.yaml]: Why is Host capitalized within the sentence?

File: api/swagger.yaml [hunk 2]
```diff
@@ -2012,28 +2012,45 @@ definitions:
       Version:
         $ref: "#/definitions/ObjectVersion"
       CreatedAt:
+        description: |
+          Date/time at which the node was added to the swarm in
```
Comment (mstanleyjones) [inline api/swagger.yaml]: It's date and time, right? Not date or time? (throughout)

File: api/swagger.yaml [hunk 3]
```diff
@@ -2012,28 +2012,45 @@ definitions:
...
+          [RFC 3339](https://www.ietf.org/rfc/rfc3339.txt) format with nano-seconds.
         type: "string"
         format: "dateTime"
         example: "2017-08-09T07:09:37.632105588Z"
       Spec:
         $ref: "#/definitions/NodeSpec"
       Description:
+        description: |
+          Description encapsulates the properties of the Node as reported by the
+          agent.
```
Comment (mstanleyjones) [inline api/swagger.yaml]: What's the "agent"?

File: api/swagger.yaml [hunk 4]
```diff
@@ -2339,6 +2363,9 @@ definitions:
   # The Swarm information for `GET /info`. It is the same as `GET /swarm`, but
   # without `JoinTokens`.
   ClusterInfo:
+    description: |
+      ClusterInfo represents info about the cluster for outputting in "info"
```
Comment (mstanleyjones) [inline api/swagger.yaml]: Ugh I wish this had not been called ClusterInfo. We don't refer to them as clusters but as swarms.

File: api/swagger.yaml [hunk 5]
```diff
@@ -2348,10 +2375,16 @@ definitions:
       Version:
         $ref: "#/definitions/ObjectVersion"
       CreatedAt:
+        description: |
+          Date/time at which the swarm cluster was initialised in
```
Comment (mstanleyjones) [inline api/swagger.yaml]: At least you can s/swarm cluster/swarm in all these descriptions.

File: api/swagger.yaml [hunk 6]
```diff
@@ -2513,10 +2546,12 @@ definitions:
           Hosts:
             type: "array"
             description: |
-              A list of hostnames/IP mappings to add to the container's `/etc/hosts` file.
-              The format of extra hosts on swarmkit is specified in:
-              http://man7.org/linux/man-pages/man5/hosts.5.html
-                IP_address canonical_hostname [aliases...]
+              A list of hostnames/IP mappings to add to the container's
+              `/etc/hosts` file. The format of extra hosts on swarmkit is
```
Comment (mstanleyjones) [inline api/swagger.yaml]: Maybe make this less Linux-specific?

File: api/swagger.yaml [hunk 7]
```diff
@@ -2199,6 +2201,33 @@ definitions:
       CertIssuerSubject: "MBMxETAPBgNVBAMTCHN3YXJtLWNh"
       CertIssuerPublicKey: "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEmT9XIw9h1qoNclv9VeHmf/Vi6/uI2vFXdBveXTpcPjqx6i9wNazchk1XWV/dKTKvSh9xyGKmiIeRcE4OiMnJ1A=="
 
+  NodeStatus:
+    description: |
+      NodeStatus represents the status of a node.
+
+      It rovides the current status of the node, as seen by the manager.
```
Comment (mstanleyjones) [inline api/swagger.yaml]: s/rovides/provides

File: api/swagger.yaml [hunk 8]
```diff
@@ -2228,6 +2230,37 @@ definitions:
...
+      Leader:
+        type: "boolean"
+        default: false
+        example: true
+      Reachability:
+        $ref: "#/definitions/Reachability"
+      Addr:
+        description: |
+          Addr provides the IP address and port at which the manager is
+          reachable. As configured through `--advertise-addr`.
```
Comment (mstanleyjones) [inline api/swagger.yaml]: s/As configured through/To configure, use the `--advertise-addr` flag

(or something like that)

File: api/swagger.yaml [hunk 9]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+          Both _tagged_ and _untagged_ (dangling) images are counted.
+        type: "integer"
+        example: 508
+      Driver:
+        description: "Name of the storage driver in use."
+        type: "string"
+        example: "overlay2"
+      DriverStatus:
+        description: |
+          Storage driver-specific information, provided as "label" / "value" pairs.
```
Comment (mstanleyjones) [inline api/swagger.yaml]: Maybe "Information specific to the storage driver"

File: api/swagger.yaml [hunk 10]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+
+          This information is provided by the storage driver, and formatted
+          for outputing as part of `docker info` on the command line.
```
Comment (mstanleyjones) [inline api/swagger.yaml]: "outputing" is not a word. Maybe "formatted in a way consistent with the output of `docker info`...." or something like that.

File: api/swagger.yaml [hunk 11]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+          - ["Backing Filesystem", "extfs"]
+          - ["Supports d_type", "true"]
+          - ["Native Overlay Diff", "true"]
+      DockerRootDir:
+        description: |
+          Root directory of persistent Docker state.
+
+          Defaults to `/var/lib/docker` on Linux, and `C:\ProgramData\docker`
+          on Windows.
+        default: "/var/lib/docker"
```
Comment (mstanleyjones) [inline api/swagger.yaml]: So the default default is Linux? ;)

File: api/swagger.yaml [hunk 12]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+        type: "string"
+        example: "/var/lib/docker"
+      SystemStatus:
+        description: |
+          Status information about this node (standalone Swarm API).
+
+          <p><br /></p>
+
+          > **Note**: The information returned in this field is only propagated
+          > by the Swarm standalone API, and is empty (`null`) in other setups.
```
Comment (mstanleyjones) [inline api/swagger.yaml]: Ehhhh.... this is interesting. Do we need to say that this doesn't include built-in swarm mode?

File: api/swagger.yaml [hunk 13]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+          - [" swarm-agent-00", "192.168.99.102:2376"]
+          - ["  └ ID", "5CT6:FBGO:RVGO:CZL4:PB2K:WCYN:2JSV:KSHH:GGFW:QOPG:6J5Q:IOZ2|192.168.99.102:2376"]
+          - ["  └ Status", "Healthy"]
+          - ["  └ Containers", "1 (1 Running, 0 Paused, 0 Stopped)"]
+          - ["  └ Reserved CPUs", "0 / 1"]
+          - ["  └ Reserved Memory", "0 B / 1.021 GiB"]
+          - ["  └ Labels", "kernelversion=4.4.74-boot2docker, operatingsystem=Boot2Docker 17.06.0-ce (TCL 7.2); HEAD : 0672754 - Thu Jun 29 00:06:31 UTC 2017, ostype=linux, provider=virtualbox, storagedriver=aufs"]
+          - ["  └ UpdatedAt", "2017-08-09T10:03:46Z"]
+          - ["  └ ServerVersion", "17.06.0-ce"]
+          - [" swarm-master", "192.168.99.101:2376"]
```
Comment (mstanleyjones) [inline api/swagger.yaml]: Don't we say manager?

File: api/swagger.yaml [hunk 14]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+        type: "string"
+        example: "https://index.docker.io/v1/"
+      RegistryConfig:
+        $ref: "#/definitions/RegistryServiceConfig"
+      GenericResources:
+        $ref: "#/definitions/GenericResources"
+      HttpProxy:
+        description: |
+          HTTP-proxy configured for the daemon. This value is obtained from the
+          [`HTTP_PROXY`](https://www.gnu.org/software/wget/manual/html_node/Proxies.html) environment variable.
```
Comment (mstanleyjones) [inline api/swagger.yaml]: Is this Linux-specific?

File: api/swagger.yaml [hunk 15]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+      ExperimentalBuild:
+        description: |
+          Indicates if experimental features are enabled on the daemon.
+        type: "boolean"
+        example: true
+      ServerVersion:
+        description: |
+          Version string of the daemon.
+
+          When connecting to a standalone Swarm API, this field returns the
```
Comment (mstanleyjones) [inline api/swagger.yaml]: You don't connect to an API, right?

File: api/swagger.yaml [hunk 16]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+
+          > **Note**: This field is only propagated when using standalone Swarm
+          > mode, and overlay networking using an external k/v store. Overlay
+          > networks with Swarm mode enabled use the built-in raft store, and
+          > this field will be empty.
+        type: "string"
+        example: "node5.corp.example.com:8000"
+      Runtimes:
+        description: |
+          List of [OCI compliant](https://github.com/opencontainers/runtime-spec)
```
Comment (mstanleyjones) [inline api/swagger.yaml]: Will people know what OCI is?

File: api/swagger.yaml [hunk 17]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+
+          The default can be overridden per-container at create time.
+        type: "string"
+        default: "runc"
+        example: "runc"
+      Swarm:
+        $ref: "#/definitions/SwarmInfo"
+      LiveRestoreEnabled:
+        description: |
+          Indicates if "live restore" is enabled.
```
Comment (mstanleyjones) [inline api/swagger.yaml]: Why is this quoted?

File: api/swagger.yaml [hunk 18]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+
+          If enabled, containers are kept running when the daemon is shutdown
+          or upon daemon start if running containers are detected.
+        type: "boolean"
+        default: false
+        example: false
+      Isolation:
+        description: |
+          Represents the isolation technology to use as a default for containers.
+          The supported values are platform specific.
```
Comment (mstanleyjones) [inline api/swagger.yaml]: platform-specific.

File: api/swagger.yaml [hunk 19]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+
+          If no isolation value is specified on daemon start, on Windows client,
```
Comment (mstanleyjones) [inline api/swagger.yaml]: The actual values should be monospaced and this could probably use some rewording in general.

File: api/swagger.yaml [hunk 20]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+          This option is currently not used on other platforms.
+        default: "default"
+        type: "string"
+        enum:
+          - "default"
+          - "hyperv"
+          - "process"
+      InitBinary:
+        description: |
+          Path to the docker-init binary.
```
Comment (mstanleyjones) [inline api/swagger.yaml]: mono space

File: api/swagger.yaml [hunk 21]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+
+          If only the binary name, not a full path is configured, the daemon
```
Comment (mstanleyjones) [inline api/swagger.yaml]: Reword. Something like "If the path is omitted, the daemon searches the `PATH` variable for the binary and uses the first result."

File: api/swagger.yaml [hunk 22]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+        example: "docker-init"
+      ContainerdCommit:
+        $ref: "#/definitions/Commit"
+      RuncCommit:
+        $ref: "#/definitions/Commit"
+      InitCommit:
+        $ref: "#/definitions/Commit"
+      SecurityOptions:
+        description: |
+          List of security features that are enabled on the daemon, such as
```
Comment (mstanleyjones) [inline api/swagger.yaml]: Is this Linux-specific?

File: api/swagger.yaml [hunk 23]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+          This configuration override this behavior, insecure communication with
+          registries whose resolved IP address is within the subnet described by
+          the CIDR syntax.
+
+          Registries can also be marked "insecure" by hostname. Those registries
+          are listed under `IndexConfigs` and have their `Secure` field set to
+          `false`.
+
+          > **Warning**: Insecure registries can be useful when running a local
+          > registry. However, because its use creates security vulnerabilities
```
Comment (mstanleyjones) [inline api/swagger.yaml]: This paragraph is weird. You start by talking about registries in the plural and then switch to singular (it). Might need some rewording.

File: api/swagger.yaml [hunk 24]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+            Secure: true
+            Official: true
+          "registry.internal.corp.example.com:3000":
+            Name: "registry.internal.corp.example.com:3000"
+            Mirrors: []
+            Secure: false
+            Official: false
+      Mirrors:
+        description: |
+          List of registries URLs that act as a mirror for the official
```
Comment (mstanleyjones) [inline api/swagger.yaml]: registry URLs

File: api/swagger.yaml [hunk 25]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+        type: "array"
+        items:
+          type: "string"
+        example:
+          - "https://hub-mirror.corp.example.com:5000/"
+          - "https://[2001:db8:a0b:12f0::1]/"
+
+  IndexInfo:
+    description:
+      IndexInfo contains information about a registry
```
Comment (mstanleyjones) [inline api/swagger.yaml]: Missing period

File: api/swagger.yaml [hunk 26]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+        type: "string"
+        example: "docker.io"
+      Mirrors:
+        description: |
+          List of mirrors, expressed as URIs.
+        type: "array"
+        items:
+          type: "string"
+        example:
+          - "https://hub-mirror.corp.example.com:5000/"
```
Comment (mstanleyjones) [inline api/swagger.yaml]: Does this work, having three different examples? Did you mean to include a `|`?

File: api/swagger.yaml [hunk 27]
```diff
@@ -3349,6 +3349,773 @@ definitions:
...
+      The runtime is invoked by the daemon via the `containerd` daemon. OCI
+      runtimes act as an interface to the Linux kernel namespaces, cgroups,
+      and SELinux.
+    type: "object"
+    properties:
+      path:
+        description: |
+          Name and, optional, path, of the OCI executable binary.
+
+          If only the binary name is configured, the daemon looks for a binary
```
Comment (mstanleyjones) [inline api/swagger.yaml]: Same rewording request as above.

**Review by mdlinville — CHANGES_REQUESTED** [review]
A few suggestions and comments. This was big, but I tried to review everything.

File: api/swagger.yaml (hunk 2 above)
Comment (thaJeztah) [inline api/swagger.yaml]: Yes it is, actually didn't realize it's ambiguous, ha! Ill update

File: api/swagger.yaml (hunk 13 above)
Comment (thaJeztah) [inline api/swagger.yaml]: Ah, yes; I had to quickly setup a legacy swarm cluster to check the responses, and copied the responses from the quickstart, which appears to use `"master"` for the manager node :smile:; https://docs.docker.com/swarm/install-w-machine/#launch-the-swarm-manager. I'll update the example output

File: api/swagger.yaml (hunk 14 above)
Comment (thaJeztah) [inline api/swagger.yaml]: I don't think it's Linux specific https://github.com/golang/go/blob/f9cf8e5ab11c7ea3f1b9fde302c0a325df020b1a/src/net/http/transport.go#L250-L300

File: api/swagger.yaml (hunk 22 above)
Comment (thaJeztah) [inline api/swagger.yaml]: Windows currently doesn't have security options, so this list would be empty; I can add a note about that

File: api/swagger.yaml (hunk 11 above)
Comment (thaJeztah) [inline api/swagger.yaml]: Let me remove the default; the default is different for each platform, and the swagger spec won't allow me expressing that

File: api/swagger.yaml (hunk 26 above)
Comment (thaJeztah) [inline api/swagger.yaml]: No this is correct; you can configure multiple mirrors for Docker Hub

**Comment by thaJeztah** [comment]
@mstanleyjones updated, PTAL

**Comment by thaJeztah** [comment]
Looks like there was an issue with Docker Hub search. Reminds me we need to rewrite that test to not rely on Hub;

```
10:16:12 ----------------------------------------------------------------------
10:16:12 FAIL: docker_cli_search_test.go:47: DockerSuite.TestSearchCmdOptions
10:16:12 
10:16:12 docker_cli_search_test.go:72:
10:16:12     c.Assert(outSearchCmdOfficialSlice, checker.HasLen, 3) // 1 header, 1 line, 1 carriage return
10:16:12 ... obtained []string = []string{"NAME      DESCRIPTION   STARS     OFFICIAL   AUTOMATED", ""}
10:16:12 ... n int = 3
10:16:12 
```

**Review by mdlinville — APPROVED** [review]

**Comment by mstanleyjones** [comment]
Approved even though it looks like there might be a few things still pending.

**Comment by thaJeztah** [comment]
Failure on `z` is not related; https://jenkins.dockerproject.org/job/Docker-PRs-s390x/5080/console

```
09:24:32 --- FAIL: TestFrequency (0.21s)
09:24:32 	splunk_test.go:891: Unexpected number of requests 7
09:24:32 time="2017-08-16T09:24:19Z" level=error msg="splunk: failed to send event - 500 Internal Server Error - " 
09:24:32 time="2017-08-16T09:24:19Z" level=error msg="splunk: failed to send event - 500 Internal Server Error - " 
09:24:32 time="2017-08-16T09:24:19Z" level=error msg="splunk: failed to send event - 500 Internal Server Error - " 
09:24:32 time="2017-08-16T09:24:19Z" level=error msg="splunk: failed to send event - 500 Internal Server Error - " 
09:24:32 time="2017-08-16T09:24:20Z" level=error msg="splunk: failed to send event - 500 Internal Server Error - " 
09:24:32 time="2017-08-16T09:24:20Z" level=error msg="splunk: failed to send event - 500 Internal Server Error - " 
09:24:32 time="2017-08-16T09:24:20Z" level=error msg="splunk: failed to send event - 500 Internal Server Error - " 
09:24:32 time="2017-08-16T09:24:20Z" level=error msg="splunk: failed to send event - 500 Internal Server Error - " 
09:24:32 time="2017-08-16T09:24:20Z" level=error msg="splunk: failed to send event - 500 Internal Server Error - " 
09:24:32 time="2017-08-16T09:24:20Z" level=error msg="Failed to send a message '{\"event\":{\"line\":\"0\",\"source\":\"stdout\",\"tag\":\"containeriid\"},\"time\":\"1502875460.655946\",\"host\":\"9e7b271ec137\"}'" 
09:24:32 time="2017-08-16T09:24:20Z" level=error msg="Failed to send a message '{\"event\":{\"line\":\"1\",\"source\":\"stdout\",\"tag\":\"containeriid\"},\"time\":\"1502875460.655974\",\"host\":\"9e7b271ec137\"}'" 
09:24:32 time="2017-08-16T09:24:20Z" level=error msg="splunk: failed to send event - 500 Internal Server Error - " 
09:24:32 time="2017-08-16T09:24:20Z" level=error msg="splunk: failed to send event - 500 Internal Server Error - " 
09:24:32 time="2017-08-16T09:24:20Z" level=error msg="Failed to send a message '{\"event\":{\"line\":\"0\",\"source\":\"stdout\",\"tag\":\"containeriid\"},\"time\":\"1502875460.718288\",\"host\":\"9e7b271ec137\"}'" 
09:24:32 time="2017-08-16T09:24:20Z" level=error msg="Failed to send a message '{\"event\":{\"line\":\"1\",\"source\":\"stdout\",\"tag\":\"containeriid\"},\"time\":\"1502875460.718302\",\"host\":\"9e7b271ec137\"}'" 
09:24:32 time="2017-08-16T09:24:20Z" level=error msg="splunk: failed to send event - 500 Internal Server Error - " 
09:24:32 time="2017-08-16T09:24:20Z" level=error msg="Failed to send a message '{\"event\":{\"line\":\"2\",\"source\":\"stdout\",\"tag\":\"containeriid\"},\"time\":\"1502875460.718303\",\"host\":\"9e7b271ec137\"}'" 
09:24:32 time="2017-08-16T09:24:20Z" level=error msg="Failed to send a message '{\"event\":{\"line\":\"3\",\"source\":\"stdout\",\"tag\":\"containeriid\"},\"time\":\"1502875460.718616\",\"host\":\"9e7b271ec137\"}'" 
09:24:32 time="2017-08-16T09:24:20Z" level=error msg="Failed to send a message '{\"event\":{\"line\":\"4\",\"source\":\"stdout\",\"tag\":\"containeriid\"},\"time\":\"1502875460.718619\",\"host\":\"9e7b271ec137\"}'" 
09:24:32 FAIL
```

All others are green; author LGTM

**Pull request merged**